src/basic/fs-util.c

```diff
@@ -435,6 +435,31 @@ int mkfifo_atomic(const char *path, mode_t mode) {
         return 0;
 }
 
+int mkfifoat_atomic(int dirfd, const char *path, mode_t mode) {
+        _cleanup_free_ char *t = NULL;
+        int r;
+
+        assert(path);
+
+        if (path_is_absolute(path))
+                return mkfifo_atomic(path, mode);
+
+        /* We're only interested in the (random) filename.  */
+        r = tempfn_random_child(NULL, NULL, &t);
+        if (r < 0)
+                return r;
+
+        if (mkfifoat(dirfd, basename(t), mode) < 0)
+                return -errno;
+
+        if (renameat(dirfd, basename(t), dirfd, path) < 0) {
+                unlink_noerrno(t);
+                return -errno;
+        }
```

Comment:
hmm, sorry for not noticing this earlier, but instead of passing NULL as path to tempfn_random_child() and then ignoring the prefix, could you instead change tempfn_random_child() to accept the empty string as path, in which case you just get the string you need, without anything prefixed? That'd be much cleaner